components/web_discovery/browser/web_discovery_service.cc

```diff
@@ -0,0 +1,117 @@
+#include "extensions/buildflags/buildflags.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/service_manager/public/cpp/interface_provider.h"
+
+namespace web_discovery {
+
+WebDiscoveryService::WebDiscoveryService(
+    PrefService* local_state,
+    PrefService* profile_prefs,
+    base::FilePath user_data_dir,
+    scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory)
+    : local_state_(local_state),
+      profile_prefs_(profile_prefs),
+      user_data_dir_(user_data_dir),
+      shared_url_loader_factory_(shared_url_loader_factory) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  if (profile_prefs_->GetBoolean(kWebDiscoveryExtensionEnabled)) {
+    profile_prefs_->ClearPref(kWebDiscoveryExtensionEnabled);
+    profile_prefs_->SetBoolean(kWebDiscoveryNativeEnabled, true);
+  }
+#endif
+
+  pref_change_registrar_.Init(profile_prefs);
+  pref_change_registrar_.Add(
+      kWebDiscoveryNativeEnabled,
+      base::BindRepeating(&WebDiscoveryService::OnEnabledChange,
+                          base::Unretained(this)));
+
+  if (profile_prefs_->GetBoolean(kWebDiscoveryNativeEnabled)) {
+    Start();
+  }
+}
+
+WebDiscoveryService::~WebDiscoveryService() = default;
+
+void WebDiscoveryService::RegisterLocalStatePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterTimePref(kPatternsRetrievalTime, {});
+}
+
+void WebDiscoveryService::RegisterProfilePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterBooleanPref(kWebDiscoveryNativeEnabled, false);
+  registry->RegisterDictionaryPref(kAnonymousCredentialsDict);
+  registry->RegisterStringPref(kCredentialRSAPrivateKey, {});
+  registry->RegisterStringPref(kCredentialRSAPublicKey, {});
+}
+
+void WebDiscoveryService::SetExtensionPrefIfNativeDisabled(
+    PrefService* profile_prefs) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  if (!base::FeatureList::IsEnabled(features::kWebDiscoveryNative) &&
+      profile_prefs->GetBoolean(kWebDiscoveryNativeEnabled)) {
+    profile_prefs->SetBoolean(kWebDiscoveryExtensionEnabled, true);
+  }
+#endif
+}
+
+void WebDiscoveryService::Start() {
+  if (!server_config_loader_) {
+    server_config_loader_ = std::make_unique<ServerConfigLoader>(
+        local_state_, user_data_dir_, shared_url_loader_factory_.get(),
+        base::BindRepeating(&WebDiscoveryService::OnConfigChange,
+                            base::Unretained(this)),
+        base::BindRepeating(&WebDiscoveryService::OnPatternsLoaded,
+                            base::Unretained(this)));
+    server_config_loader_->LoadConfigs();
+  }
+  if (!credential_manager_) {
+    credential_manager_ = std::make_unique<CredentialManager>(
+        profile_prefs_, shared_url_loader_factory_.get(),
+        server_config_loader_.get());
+  }
+}
+
+void WebDiscoveryService::Stop() {
+  server_config_loader_ = nullptr;
+  credential_manager_ = nullptr;
+
+  profile_prefs_->ClearPref(kWebDiscoveryNativeEnabled);
```

Comment:
I don't think `Stop` should clear `kWebDiscoveryNativeEnabled`. Other prefs cleanup is understandable.

Reply:
using `Start` and `Stop` seems confusing to me. Shouldn't this be `Enable` and `Disable`? As mentioned above having separate prefs for native and extension seems unnecessarily complex and confusing.

Reply:
i think the current names are appropriate. `Start` also gets called on profile startup, so `Enable` doesn't seem right.